src/pokeathlon/pokeathlon_course.c

```diff
@@ -87,3 +101,152 @@ BOOL PokeathlonCourse_Init(OverlayManager *manager, int *state) {
 
     return TRUE;
 }
+
+BOOL PokeathlonCourse_Main(OverlayManager *manager, int *state) {
+    PokeathlonCourseData *data;
+    PokeathlonCourseState *courseState;
+
+    data = OverlayManager_GetData(manager);
+    courseState = &data->courseState;
+
+    // Frame counter system: increments frameTimer each frame, and frameCounter every 0x708 frames (up to max 0xEA5F)
+    if (data->counterEnabled != 0) {
+        data->frameTimer++;
+        if (data->frameTimer >= 0x708) {
+            if (data->frameCounter < 0xEA5F) {
+                data->frameCounter++;
+            }
+            data->frameTimer = 0;
+        }
+    }
+
+    // Main state machine
+    switch (courseState->mainState) {
+    case 0: // Idle state - waiting for transitions
+        if (ov96_021E5C2C(data)) {
+            return TRUE;
+        }
+
+        // Handle exit flag
+        if (courseState->exitFlag != 0) {
+            PokeathlonCourseArgs *args;
+            args = (PokeathlonCourseArgs *)courseState->argsPtr;
+            args->modeBytes[2] = args->modeBytes[3];
+            args->modeBytes[1] = 0;
+            courseState->exitFlag = 0;
+        }
+
+        // Check for state transitions
+        if (courseState->transitionType != 0) {
+            if (courseState->transitionType == 0x10) { // Special exit transition
+                courseState->mainState = 3;            // Go to state 3
+            } else {
+                courseState->mainState = 1; // Go to state 1
+            }
+        }
+        break;
+
+    case 1: // Start transition
+        sub_02037AC0((u8)courseState->transitionType);
+        courseState->mainState = 2;
+        // Fallthrough to case 2
+
+    case 2: // Wait for transition to complete
+        if (sub_02037B38((u8)courseState->transitionType)) {
+            courseState->mainState = 0;      // Return to idle
+            courseState->transitionType = 0; // Clear transition type
+        }
+        break;
+
+    case 3: // Start exit transition (special case for transition type 0x10)
+        if (courseState->transitionType != 0x10) {
+            GF_AssertFail();
+        }
+        sub_02037AC0((u8)courseState->transitionType);
```

Comment:
for transitionType
would it work as a u8 to avoid all the casting or does that not match?
casting might not even be necessary if sub_02037AC0 has the right parameter type